### PR TITLE
Extract ScreenshotConfiguration from ScreenshotRule

### PR DESCRIPTION
**Pull request by DanielJette**
### What does this change accomplish?

This is the first of many PRs to break down the monolithic `ScreenshotRule` into a collection of composable, reusable and extensible elements. The ultimate goal is to allow Testify to interoperate with a wider variety of testing methodologies such as JUnit5 and Test Scenarios.

### How have you achieved it?

In this PR, I have collected all of the _configuration_ related properties from `ScreenshotRule` and extracted them to a new interface, `ScreenshotConfigurationInterface`. I have then employed the [delegation pattern](https://en.wikipedia.org/wiki/Delegation_pattern) to compose `ScreenshotRule` from `ScreenshotConfiguration`. The primary outcome from this change is to reduce the size of the `ScreenshotRule` class and begin the process of more narrowly defining the role and purpose of each aspect of the class. Eventually, all of `ScreenshotRule` will be replaced with reusable, composable objects.

image

### Tophat instructions

There are no expected changes to the public API in this PR. All tests should pass without modification.

### Notice

This change must keep `main` in a shippable state; **it may be shipped without further notice**.

### CODE REVIEW

File: Library/src/main/java/com/shopify/testify/core/ScreenshotConfigurationInterface.kt [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+import com.shopify.testify.internal.modification.FocusModification
+import com.shopify.testify.internal.modification.HideCursorViewModification
+import com.shopify.testify.internal.modification.HidePasswordViewModification
+import com.shopify.testify.internal.modification.HideScrollbarsViewModification
+import com.shopify.testify.internal.modification.HideTextSuggestionsViewModification
+import com.shopify.testify.internal.modification.SoftwareRenderViewModification
+import com.shopify.testify.report.Reporter
+import java.util.Locale
+
+typealias ViewModification = (rootView: ViewGroup) -> Unit
```
Comment (lwise) [inline Library/src/main/java/com/shopify/testify/core/ScreenshotConfigurationInterface.kt:41]: I know for the scope of this PR this is just being moved, but I'm curious as to why we have this `ViewModification` typealias when we also have the `com.shopify.testify.internal.modification.ViewModification` class, and they appear to do very similar things (although I guess some are more "first class" in terms of having autocomplete/api support). I guess my other question is why we only allow one to be set at a time.

**Review by lwise — COMMENTED** [review]
Looks great; commit structure was really nice to follow so thanks 😄

File: Library/src/main/java/com/shopify/testify/ScreenshotRule.kt [hunk 2]
```diff
@@ -103,46 +96,15 @@ open class ScreenshotRule<T : Activity> @JvmOverloads constructor(
     initialTouchMode: Boolean = false,
     protected val launchActivity: Boolean = true,
     enableReporter: Boolean = false
-) : ActivityTestRule<T>(activityClass, initialTouchMode, launchActivity), TestRule {
+) : ActivityTestRule<T>(activityClass, initialTouchMode, launchActivity),
+    TestRule,
+    ScreenshotConfigurationInterface by ScreenshotConfiguration() {
```
[the remark this answers is not in the document]
Comment (DanielJette) [reply Library/src/main/java/com/shopify/testify/ScreenshotRule.kt:101]: I see the ScreenshotConfiguration as an encapsulation of all the "optional bits" that testify can use to configure the way it handles screenshots. What I want eventually is for the ScreenshotRule to be a composition of a bunch of helper classes and only layer on the Android Test Rule interface.
By having the configuration in a separate class, I can reuse it with the other implementations (like a junit5 extension or a test scenario)

File: Library/src/main/java/com/shopify/testify/core/ScreenshotConfigurationInterface.kt (hunk 1 above)
Comment (DanielJette) [reply Library/src/main/java/com/shopify/testify/core/ScreenshotConfigurationInterface.kt:41]: Yeah, this is an ugly holdover from the java & junit3 days. Basically, we had an internal set of functions that were all named "setBlahViewModification" and then also the public `setViewModification` method that's used in the test code. They were initially intended to be the same thing, but they kind of evolved a bit differently when we started refactoring to kotlin. In retrospect, we should never have used the same name for both things, or we should have made them literally the same. It's kind of techdebt, I guess.

**Pull request closed without merging**